build-system/common/ci.js

```diff
@@ -15,121 +15,231 @@
-  isGithubActionsPushBuild,
-} = require('./github-actions');
-const {
-  isTravisBuild,
-  isTravisPullRequestBuild,
-  isTravisPushBuild,
-  travisPullRequestBranch,
-  travisPullRequestSha,
-} = require('./travis');
-
-/**
- * @fileoverview Provides functions that extract various kinds of CI state.
- */
-
-/**
- * Maps generic CI functions to those that must be run on the current service.
- */
-const serviceFunctionMap = isTravisBuild()
-  ? {
-      'ciPullRequestBranch': travisPullRequestBranch,
-      'ciPullRequestSha': travisPullRequestSha,
-      'isPullRequestBuild': isTravisPullRequestBuild,
-      'isPushBuild': isTravisPushBuild,
-    }
-  : isGithubActionsBuild()
-  ? {
-      'ciPullRequestBranch': githubActionsPullRequestBranch,
-      'ciPullRequestSha': githubActionsPullRequestSha,
-      'isPullRequestBuild': isGithubActionsPullRequestBuild,
-      'isPushBuild': isGithubActionsPushBuild,
-    }
-  : isCircleciBuild()
-  ? {
-      'ciPullRequestBranch': circleciPullRequestBranch,
-      'ciPullRequestSha': circleciPullRequestSha,
-      'isPullRequestBuild': isCircleciPullRequestBuild,
-      'isPushBuild': isCircleciPushBuild,
-    }
-  : {
-      'ciPullRequestBranch': () => '',
-      'ciPullRequestSha': () => '',
-      'isPullRequestBuild': () => false,
-      'isPushBuild': () => false,
-    };
+/**
+ * Shorthand to extract an environment variable.
+ * @param {string} key
+ * @return {string|undefined}
+ */
+function env(key) {
+  return process.env[key];
+}
 
 /**
  * Returns true if this is a CI build.
  * @return {boolean}
  */
 function isCiBuild() {
-  // Travis: https://docs.travis-ci.com/user/environment-variables/#default-environment-variables
-  // GitHub Actions: https://docs.github.com/en/free-pro-team@latest/actions/reference/environment-variables#default-environment-variables
-  // CircleCI: https://circleci.com/docs/2.0/env-vars/#built-in-environment-variables
-  return !!process.env.CI;
+  return !!env('CI');
 }
 
+/**
+ * Returns true if this is a Travis build.
+ * @return {boolean}
+ */
+function isTravisBuild() {
+  return !!env('TRAVIS');
+}
+
+/**
+ * Returns true if this is a GitHub Actions build.
+ * @return {boolean}
+ */
+function isGithubActionsBuild() {
+  return !!env('GITHUB_ACTIONS');
+}
+
+/**
+ * Returns true if this is a CircleCI build.
+ * @return {boolean}
+ */
+function isCircleciBuild() {
+  return !!env('CIRCLECI');
+}
+
+/**
+ * Constants for reduced code size.
+ */
+const isTravis = isTravisBuild();
+const isGithubActions = isGithubActionsBuild();
+const isCircleci = isCircleciBuild();
+
 /**
  * Returns true if this is a PR build.
  * @return {boolean}
  */
 function isPullRequestBuild() {
-  return serviceFunctionMap['isPullRequestBuild']();
+  return isTravis
+    ? env('TRAVIS_EVENT_TYPE') === 'pull_request'
+    : isGithubActions
+    ? env('GITHUB_EVENT_NAME') === 'pull_request'
+    : isCircleci
+    ? !!env('CIRCLE_PULL_REQUEST')
+    : false;
```

Comment:
I personally find multiple `?:` to be extremely difficult to read. How about rewriting this as:
```javascript
return (isTravis && env('TRAVIS_EVENT_TYPE') === 'pull_request') ||
  (isGithubActions && env('GITHUB_EVENT_NAME') === 'pull_request') ||
  (isCircleci && !!env('CIRCLE_PULL_REQUEST');
```

Reply:
Your suggestion will only work for boolean return values. The only other alternative seems to be this:

```js
  if (isTravis) {
    return env('TRAVIS_EVENT_TYPE') === 'pull_request';
  }
  if (isGithubActions) {
    return env('GITHUB_EVENT_NAME') === 'pull_request';
  }
  if (isCircleci) {
    return !!env('CIRCLE_PULL_REQUEST');
  }
  return false;
```

To be honest, I hemmed and hawed over this. Chained (not nested!) ternary was more readable to me than the ~30% more verbose `if` ladder. This (admittedly opinionated) article eventually convinced me: https://medium.com/javascript-scene/nested-ternaries-are-great-361bddd0f340

Reply:
In my opinion, having a separate module, class, or object for each type offered a better encapsulation of logic and a cleaner syntax than the ternary chains

Reply:
Early versions of this refactor attempted to use a separate module per service. The problems were: 1) there is no clean 1:1 mapping from each piece of state to each service, 2) the extra level of indirection obfuscated service-level differences, and 3) the sub-modules mostly contained boilerplate functions / getters that added to the maintenance burden.

At this point, unless there is strong objection, I'd like to merge this PR in the interest of making progress on the rest of the updates needed for this migration.

Back at y'all for final comment :)